keps/sig-node/3960-pod-lifecycle-sleep-action/README.md

```diff
@@ -0,0 +1,376 @@
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes the addition of a new sleep action for the PreStop lifecycle hook in Kubernetes, allowing containers to pause for a specified duration before termination. This enhancement aims to provide a more straightforward way to manage graceful shutdowns and improve the overall lifecycle management of containers.
+
+## Motivation
+
+Currently, Kubernetes supports two types of actions for PreStop hooks: exec and httpGet. Although these actions offer flexibility, they often require additional scripting or custom solutions to achieve a simple sleep functionality. A built-in sleep action would provide a more user-friendly and native solution for scenarios where a container needs to pause before shutting down, such as:
+
+- Ensuring that the container gracefully releases resources and connections.
+- Allowing a smooth transition in load balancers or service meshes.
+- Providing a buffer period for external monitoring and alerting systems.
+
+### Goals
+
+- Allow containers to perform cleanup or shutdown actions before being terminated, by sleeping for a specified duration in the preStop hook.
+- Improve the overall reliability and availability of Kubernetes applications by providing a way for containers to gracefully terminate.
+
+### Non-Goals
+
+- This KEP does not aim to replace other Kubernetes features that can be used to perform cleanup actions, such as init containers or sidecar containers.
+- This KEP does not aim to provide a way to pause or delay pod termination indefinitely.
+
+## Proposal
+
+We propose adding a new sleep action for the PreStop hook, which will pause the container for a specified duration before termination. The API changes will include the following:
+
+- Extending the LifecycleHandler object to support a new Sleep field.
+- Adding a SleepAction object that includes a Duration field to specify the sleep period in seconds.
+
+### User Stories (Optional)
```

Comment:
Hmm - TBH the user stories aren't really convincing to me:
1) the sleep often doesn't guarantee that resources will be released (if the application is badly written - and if it's well written it usually is easy to expose the httpGet for checking that)
2) sleep doesn't guarantee being remove from LB - for many setups this highly depends on many different things
3) external monitoring/alerting - that's closest, but wouldn't it be better to expose the signal from those solutions and let it be used as a PreStop hook instead?

Reply:
Hmmm, then maybe just one user story:
As a Kubernetes user, I want to configure my container to sleep for a specific duration after receiving sigterm, but I don't want to introduce a sleep binary to my image. Like originally posted in [#114465](https://github.com/kubernetes/kubernetes/issues/114465)

Reply:
@wojtek-t I agree that "sleep a while" is horrible BUT it's accepted practice when running things like nginx which do the wrong thing upon SIGTERM:

https://www.google.com/search?q=kubernetes+nginx+prestop

So the real user-story here is "Let users do things they already do, more safely" (without needing a sleep binary in their image).

Reply:
If the goal is to achieve synchronization with something else (endpoints cleanup in this case) wouldn't it be better to make it possible to do this explicitly somehow, rather than promote this ugly hack that is not 100% reliable anyway?

I hear there are similar issues with some NetworkPolicy-respecting CNI plugins which remove terminating pods from allowlists before the clients stop sending traffic, leading to apparent policy violations.

These sound like more fundamental issues with pod lifecycle 🤔

Reply:
@porridge please file an issue if you believe the API is not working. As mentioned before - a lot of goodness can be achieved using this new sleep prestop hook. Further improvements in pod lifecycle are also not blocked by this API.

Reply:
@porridge I'll be first to agree that the relationship between pod lifecycle and network is under-developed.  That said, we don't have the mechanisms needed to fully interlock these concepts, and doing so would be very complex WRT things like CRD-based workload controllers.  It would require a pretty heavy re-imagining of the pod lifecycle and of what it means to delete an API object.  It's possible that we can get there (there is some work afoot to make pod-removal more coordinated) but it's "in addition to", not "instead of", this KEP.